### DISCUSSION TITLE
New API hook for financial payments

### DISCUSSION THREAD

**Discussion by DanStevens** [OpenRCT2/OpenRCT2#17523] [Feature requests]
I'd like to contribute a new API hook for financial payments named "park.finance.payment". This hook would be triggered whenever payment occurs via the [`financial_payment`](https://github.com/OpenRCT2/OpenRCT2/blob/5fe4c565067ddef2f7980b37c025ab1bd8c3c467/src/openrct2/management/Finance.cpp#L87) function is called and would provide the ExpenditureType and amount as arguments. If the plugin modifies the amount on the argument, the new value is supplied back to the game host and the new value is assumed when updating the game state. In simple terms, whenever the player or game does something that changes the cash balance, the hook is triggered.

A use case for this API that I have in mind is a plugin that can be used to increase or decrease difference expenses by a user defined factor (e.g. x2 or x0.5) in order to make the financial aspect of the game harder or easier.

While I'm more of a C# programmer than a C++ programmer, I've taken a look at the code and I'm confident enough to give this a go. I've been reviewing the code that implements the "ride.ratings.calculate" hook, which looks very similar in principle and it looks like it supports feedback of new ride stats to the game. In fact, I have some success in getting this partially working:

![OpenRCT2-ParkFinancePaymentApiHookTest](https://user-images.githubusercontent.com/1074238/177880775-3a4d1467-d47d-403e-bdf4-c575dfc0e045.jpg)

The diff below for openrct2.d.ts may help illustrate what I have in mind. I welcome any thoughts, feedback or pointers from developers or users on this. I hope it would be a useful contribution to others as well as myself.

```diff
diff --git "a/distribution/openrct2.d.ts" "b/distribution/openrct2.d.ts"
index 61258556a..f9ab528cc 100644
--- "a/distribution/openrct2.d.ts"
+++ "b/distribution/openrct2.d.ts"
@@ -299,6 +299,7 @@ declare global {
         subscribe(hook: "vehicle.crash", callback: (e: VehicleCrashArgs) => void): IDisposable;
         subscribe(hook: "map.save", callback: () => void): IDisposable;
         subscribe(hook: "map.change", callback: () => void): IDisposable;
+        subscribe(hook: "park.finance.payment", callback: (e: PaymentEventArgs) => void): IDisposable;
 
         /**
          * Can only be used in intransient plugins.
@@ -412,7 +413,8 @@ declare global {
         "interval.tick" | "interval.day" |
         "network.chat" | "network.action" | "network.join" | "network.leave" |
         "ride.ratings.calculate" | "action.location" | "vehicle.crash" |
-        "map.change" | "map.changed" | "map.save";
+        "map.change" | "map.changed" | "map.save" |
+        "park.finance.payment";
 
     type ExpenditureType =
         "ride_construction" |
@@ -574,6 +576,19 @@ declare global {
         readonly crashIntoType: VehicleCrashIntoType;
     }
 
+    /**
+     * A financial payment event, provided by the "park.financial.payment" hook
+     */
+    interface PaymentEventArgs {
+        /** The type of the payment */
+        readonly type: ExpenditureType
+        /**
+         * The amount of the payment, which is 100 times the amount display ingame e.g $3.50 is 35.
+         * Income i.e entrance tickets, ride tickets and food/drink sales, are negative
+         */
+        amount: number;
+    }
+
     /**
      * APIs for the in-game date.
      */

```

**Comment by Die4Ever** [comment]
Thanks for doing this, it would be great for my Randomizer plugin!

`* The amount of the payment, which is 100 times the amount display ingame e.g $3.50 is 35.`

you mean 10x not 100x?

**Reply by DanStevens** [reply]
Yes, it should be 10 times, not 100. Thanks for pointing that out.

**Comment by duncanspumpkin** [comment]
I'm not sure that would be the best place to hook this. The expenditure type isn't going to give you much fine grained control.

**Reply by DanStevens** [reply]
I think the `financial_payment` function is favorable as it's a single point of entry for all financial transactions, meaning the hook can be implemented with a single code block in one location (excluding other changes for various metadata). I'm sure there will be some limitations. For example, it would be nice to include the ride ID for transaction related to a ride. While this data isn't provided to the financial_payment function, it could probably be provided but it take some work. Probably something for a later iteration. I think the API could still have uses even without this.

I'm open to suggestions for alternative implementations. Anyone who wants to follow my progress can refer to my fork: https://github.com/DanStevens/OpenRCT2/tree/finance_payment_api_hook

**Reply by duncanspumpkin** [reply]
> 

I don't want this simple function muddied with ancillary parameters just for plugins. I was sort of thinking a hook in the staff salary calc, ride maintenance costs calc, etc.

**Comment by DanStevens** [comment]
So I have this working, as shown in the image below. I created a simple proof on concept plugin that sets every payment to zero and the park balance never changes, as expected. Also the Financial Summary window is empty, as I'd expect. I'll point out that various UI elements in the screenshot don't reflect the fact that now all payments are zero. This is to be expected as the amount is changed by the plugin after the payment is generated. This is a limitation to be aware of for plugin developers though. I can see the potential for other hooks, such as when ride running costs are calculated.

![image](https://user-images.githubusercontent.com/1074238/177998547-db36689f-396d-479a-ad40-25f27ad36f98.png)
